### PR TITLE
Feature/redis

### PR DESCRIPTION
**Pull request by wojdanowski**
- Cache user tokens in redis,
- Added logout route

### CODE REVIEW

File: controllers/authController.js [hunk 1]
```diff
@@ -74,6 +77,29 @@ exports.login = catchAsync(async (req, res, next) => {
...
+	let token;
+	if (req.headers.authorization) {
+		token = req.headers.authorization.split(' ')[1];
+	} else if (req.body.token) {
+		token = req.body.token;
+	} else if (req.cookies.jwt) {
+		token = req.cookies.jwt;
+	}
+
+	if (!token)
```
Comment (MoodupDev) [inline controllers/authController.js]: Always use curly brackets `{}`, even for single lines. This style often leads to uncaught errors and strange behaviors (it's easy to make a mistake here).

File: redis.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+let redis_url = process.env.REDIS_URL;
+let redis_port = process.env.REDIS_PORT;
+
+if (process.env.NODE_ENV === 'development') {
+	redis_url = '127.0.0.1';
+	redis_port = 6379;
+}
+const redisClient = redis.createClient(redis_port, redis_url);
+redisClient.on('connect', function () {
```
Comment (MoodupDev) [inline redis.js]: not an issue, but I think you can use arrow function here: `function () {...}` -> `() -> {...}`

File: redis.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
+const redis = require('redis');
+const { promisify } = require('util');
+
+let redis_url = process.env.REDIS_URL;
+let redis_port = process.env.REDIS_PORT;
+
+if (process.env.NODE_ENV === 'development') {
```
Comment (MoodupDev) [inline redis.js]: Your .env on your local machine file should have all necessary env variables for development environment.

File: redis.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+	redis_url = '127.0.0.1';
+	redis_port = 6379;
+}
+const redisClient = redis.createClient(redis_port, redis_url);
+redisClient.on('connect', function () {
+	console.log('Redis connected');
+});
+
+const getAsync = promisify(redisClient.get).bind(redisClient);
```
Comment (MoodupDev) [inline redis.js]: I would change names of these to `getCache`, `setCache`, `deleteCache`

**Review by MoodupDev — CHANGES_REQUESTED** [review]

File: controllers/authController.js [hunk 5]
```diff
@@ -74,6 +77,29 @@ exports.login = catchAsync(async (req, res, next) => {
 	createSendToken(user, StatusCodes.OK, res);
 });
 
+exports.logout = catchAsync(async (req, res, next) => {
+	let token;
+	if (req.headers.authorization) {
+		token = req.headers.authorization.split(' ')[1];
```
Comment (arkus7) [inline controllers/authController.js]: I guess you could also check if the prefix is `bearer`/`Bearer` but it's not necessary

File: controllers/authController.js [hunk 6]
```diff
@@ -74,6 +77,29 @@ exports.login = catchAsync(async (req, res, next) => {
...
+	} else if (req.cookies.jwt) {
+		token = req.cookies.jwt;
+	}
+
+	if (!token)
+		return next(new AppError('No token provided', StatusCodes.BAD_REQUEST));
+
+	const deletedRecords = await delAsync(token);
+
+	if (!deletedRecords)
```
Comment (arkus7) [inline controllers/authController.js]: Same here, add curly brackets

File: controllers/authController.js [hunk 7]
```diff
@@ -74,6 +77,29 @@ exports.login = catchAsync(async (req, res, next) => {
...
+		return next(new AppError('User already logout', StatusCodes.NOT_FOUND));
```
Comment (arkus7) [inline controllers/authController.js]: I'm not sure if you should throw not found here. 
As a user, I wanted to be logged out, if I'm already logged out this satisfied my request, so I guess you could respond with `204 No content` or just `200 Ok`

What do you think @MoodupDev

File: passport/strategies.js [hunk 8]
```diff
@@ -2,17 +2,27 @@ const BearerStrategy = require('passport-http-bearer').Strategy;
...
 const { promisify } = require('util');
 const User = require('./../models/userModel');
+const { getAsync, setAsync } = require('./../redis');
 
 const authorizeToken = async (req, accessToken, callback) => {
 	try {
-		// 2) Verification of token
+		// Verify if token is in cache
+		let currentUser;
+		const cachedUser = await getAsync(accessToken);
```
Comment (arkus7) [inline passport/strategies.js]: For how long is the user token cached? Is there a mechanism checking that the token hasn't expired?

File: passport/strategies.js [hunk 9]
```diff
@@ -22,6 +32,13 @@ const authorizeToken = async (req, accessToken, callback) => {
 			);
 		}
 
+		// Cache user in redis and send response to client
+		await setAsync(
+			accessToken,
+			JSON.stringify(currentUser),
+			'EX',
+			process.env.REDIS_EXPIRES_IN
```
Comment (arkus7) [inline passport/strategies.js:40]: Ok, I see the logic here, you can ignore above comment :P 
Make sure that `REDIS_EXPIRES_IN` is not longer than the expiry time of access token.

File: redis.js (hunk 4 above)
Comment (arkus7) [reply redis.js]: Maybe even `getFromCache` `saveToCache`, `deleteFromCache`

**Review by arkus7 — APPROVED** [review]

File: controllers/authController.js (hunk 7 above)
Comment (MoodupDev) [reply controllers/authController.js]: @arkus7 yes I agree

File: passport/strategies.js (hunk 8 above)
Comment (wojdanowski) [reply passport/strategies.js:11]: Token is cached for 60 min.
In authorizeToken() (strategies.js) there is jwt.verify(token,...) so expired token will be rejected.

File: passport/strategies.js (hunk 9 above)
Comment (wojdanowski) [reply passport/strategies.js:40]: Token is cached for 60 min.
In authorizeToken() (strategies.js) there is jwt.verify(token,...) so expired token will be rejected.

File: controllers/authController.js [hunk 10]
```diff
@@ -80,20 +82,27 @@ exports.login = catchAsync(async (req, res, next) => {
...
+		);
+	}
 
-	const deletedRecords = await delAsync(token);
+	const deletedRecords = await delFromCache(token);
 
-	if (!deletedRecords)
-		return next(new AppError('User already logout', StatusCodes.NOT_FOUND));
+	if (!deletedRecords) {
+		return next(new AppError('User already logout', StatusCodes.OK));
```
Comment (arkus7) [inline controllers/authController.js]: Not sure whether this should be wrapped up in `AppError` as this isn't an error anymore :P

**Review by arkus7 — APPROVED** [review]

**Pull request merged**